### ISSUE TITLE
Change Request: Support passing `RuleDefinition` type to the `RuleTester.run()` method

### ISSUE THREAD

**Issue by lumirlumir** [eslint/eslint#19633]
### ESLint version

v9.24.0

### What problem do you want to solve?

Hello,

Currently, the `RuleTester.run()` method only accepts the `Rule.RuleModule` type as a valid parameter:

```js
// ...

	run(
		name: string,
		rule: Rule.RuleModule,
		tests: {
			valid: Array<string | RuleTester.ValidTestCase>;
			invalid: RuleTester.InvalidTestCase[];
		},
	): void;

// ...
```

However, `Rule.RuleModule` is specific to JavaScript rules and is not a general-purpose type for rules.

This limitation prevents me from using custom rule definitions such as `MarkdownRuleDefinition` or `JSONRuleDefintion` with `RuleTester`.

So, I’d like to propose extending support to accept a more general `RuleDefinition` type as a valid parameter for `RuleTester.run()`.

### What do you think is the correct solution?

I can think of two possible approaches, as outlined below:

```diff
	run(
		name: string,
-		rule: Rule.RuleModule,
+               rule: Rule.RuleModule | RuleDefinition
		tests: {
			valid: Array<string | RuleTester.ValidTestCase>;
			invalid: RuleTester.InvalidTestCase[];
		},
	): void;
```

or

```diff
	run(
		name: string,
-		rule: Rule.RuleModule,
+               rule: RuleDefinition
		tests: {
			valid: Array<string | RuleTester.ValidTestCase>;
			invalid: RuleTester.InvalidTestCase[];
		},
	): void;
```

I believe `RuleDefinition` is a more general type, which allows rule definitions for other languages to be used as well.

### Participation

- [x] I am willing to submit a pull request for this change.

**Comment by fasttime** [comment]
I agree that `RuleTester.run()` should be switched to accept a `rule` argument of type `RuleDefinition`. I don't think it is necessary to specify the union `Rule.RuleModule | RuleDefinition` for that type, as `RuleDefinition` alone should accept a `Rule.RuleModule` argument. We can keep [the existing unit test](https://github.com/eslint/eslint/blob/90228e5d57672579cf82bede29880532c2cb8ca9/tests/lib/types/types.test.ts#L1913) to ensure that this is the case, and add a new one for the more generic `RuleDefinition` type. A pull request is welcome.
